DO NOT MERGE: [stdlib] CharacterView not own SliceType, and drop Unicode 8 tries (finally)

milseman opened the pull request:
<!-- What's in this pull request? -->

[stdlib] String.CharacterView.SubSequence is Substring.CharacterView
    
This changes String.CharacterView from being its own slice type to
instead use Substring.CharacterView. In the process, this updates
Substring.CharacterView to be RangeReplaceable.

This aligns with the goal of introducing Substring as a separate type
to avoid accidental memory "leaks".

---

[stdlib] Drop Unicode 8 tries in stdlib (entirely)
    
This drops the last vestage of Unicode 8 tries from the standard
library. Switches everything over to use ICU.

milseman committed:
[stdlib] String.CharacterView.SubSequence is Substring.CharacterView

This changes String.CharacterView from being its own slice type to
instead use Substring.CharacterView. In the process, this updates
Substring.CharacterView to be RangeReplaceable.

This aligns with the goal of introducing Substring as a separate type
to avoid accidental memory "leaks".

milseman committed:
[stdlib] Drop Unicode 8 tries in stdlib (entirely)

This drops the last vestage of Unicode 8 tries from the standard
library. Switches everything over to use ICU.

milseman commented:
@swift-ci please test

milseman commented:
This failed due to infrastructure problems, but didn't report back.

milseman commented:
@swift-ci please test

shahmishal commented:
@milseman If you see it again can you please send me an email with Job URL. Thanks!

milseman commented:
Will do!

milseman commented:
So yeah... this is totally a source break for Swift 4 code. I'll need to do some `@available` hacky judo.

moiseev commented on stdlib/public/core/Substring.swift.gyb:
```diff
@@ -384,7 +384,7 @@ extension StringProtocol {
 % ]:
 %   View = ViewPrefix + 'View'
-%   RangeReplaceable = 'RangeReplaceable' if property == 'unicodeScalars' else ''
+%   RangeReplaceable = 'RangeReplaceable' if property == 'unicodeScalars' or property == 'characters' else ''
```
Minor nitpick: `if property in ["unicodeScalars", "characters"]`.

moiseev reviewed (commented):
LGTM

natecook1000 commented on stdlib/public/core/StringCharacterView.swift:
```diff
@@ -739,6 +741,12 @@ extension String.CharacterView {
   /// - Complexity: O(*n*) if the underlying string is bridged from
   ///   Objective-C, where *n* is the length of the string; otherwise, O(1).
+  @available(swift, introduced: 4.0)
+  public subscript(bounds: Range<Index>) -> Substring.CharacterView {
```
What's the story around substrings and bridging? Do we automatically bridge an objc string when it's substring'd, or can a substring be `NSString`-backed?

moiseev replied:
Right now when the bridging happens, a new `String` gets created from a substring [here](https://github.com/apple/swift/blob/41b00c57ab90c176e9d24098f839b315a2812a50/stdlib/public/SDK/Foundation/String.swift#L63).

milseman committed:
[stdlib] Availability checking for String.*View.popFirst

Collection's popFirst is only present when the Collection is its own
SubSequence type. String and String's views are no longer their own
SubSequenes, so popFirst is no longer present. Unfortunately, this
breaks code in swift-version 3 and it also gives a terrible diagnostic
to users. This change introduces an implementation for swift-version 3
mode and better diagnostics for Swift 4 code. Tests included.

moiseev commented on test/stdlib/StringCompatibility.swift:
```diff
@@ -356,5 +356,20 @@ Tests.test("LosslessStringConvertible/force unwrap/\(swift)") {
 #endif
 
+#if !swift(>=4)
+Tests.test("popFirst") {
```
Just a little comment, when the test is run in both Swift 3 and Swift 4 modes, it is quite useful to add a swift version in a test name. Like the following.

```swift
Tests.test("LosslessStringConvertible/force unwrap/\(swift)") {
```

It is not particularly important for this test right no as it only exists in Swift 3 world, but refactoring happens...

milseman commented:
`@available` judo is insufficient, as it cannot version associated types. This leaves us in the situation where either: we leak, we break source, or we deprecate CharacterView asap (post 4.0)

I'll split this PR into multiple ones to tackle the other issues unrelated to String.CharacterView, and mock up what the third option is like.
